From e7be1c5e84e4050fdc2ec9f4fdb14d5759b1ac25 Mon Sep 17 00:00:00 2001
From: Kirill Kapranov <kirill.kapranov@compulab.co.il>
Date: Fri, 25 Oct 2024 23:29:30 +0300
Subject: [PATCH 126/155] ucm-imx8m-mini: Introduce fdt_file autoselection

In order to support the PCB revision 1.4, equipped with the WiFi module EL2 by
Murata, instead of WB5+ by Laird, introdused the autoselection of fdt file to be
loaded. The selection is based on the presence of the sub-string 'WBM' in the
product option line (M for Murata).
A non-empty string, assigned to the u-boot environment variable 'fdt_file'
overrides the autoselection.

Signed-off-by: Kirill Kapranov <kirill.kapranov@compulab.co.il>
---
 board/compulab/ucm-imx8m-mini/Kconfig         | 12 ++++++-
 .../compulab/ucm-imx8m-mini/ucm-imx8m-mini.c  | 36 +++++++++++++++++++
 configs/cl-imx8m-mini_defconfig               |  1 +
 configs/ucm-imx8m-mini.config                 |  1 -
 4 files changed, 48 insertions(+), 2 deletions(-)

diff --git a/board/compulab/ucm-imx8m-mini/Kconfig b/board/compulab/ucm-imx8m-mini/Kconfig
index 4e33ffcdc02..cd7f88e902b 100644
--- a/board/compulab/ucm-imx8m-mini/Kconfig
+++ b/board/compulab/ucm-imx8m-mini/Kconfig
@@ -12,10 +12,20 @@ config SYS_CONFIG_NAME
 config DEFAULT_DTB
 	depends on TARGET_UCM_IMX8M_MINI
 	default "ucm-imx8m-mini.dtb"
-
+# Automatic selection of fdt_file requires the variable is empty
 config DEFAULT_FDT
+	string
+	depends on TARGET_UCM_IMX8M_MINI
+	default ""
+# fdt_file for specific WiFi/BT modules:
+config DEFAULT_FDT_WB5 # WB5 by Laird
 	string
 	depends on TARGET_UCM_IMX8M_MINI
 	default "ucm-imx8m-mini.dtb"
 
+config DEFAULT_FDT_2EL #2EL by Murata
+	string
+	depends on TARGET_UCM_IMX8M_MINI
+	default "ucm-imx8m-mini-rev1.4.dtb"
+
 endif
diff --git a/board/compulab/ucm-imx8m-mini/ucm-imx8m-mini.c b/board/compulab/ucm-imx8m-mini/ucm-imx8m-mini.c
index 6cafc259e9e..6780bcdd24f 100644
--- a/board/compulab/ucm-imx8m-mini/ucm-imx8m-mini.c
+++ b/board/compulab/ucm-imx8m-mini/ucm-imx8m-mini.c
@@ -26,5 +26,41 @@
 #include <usb.h>
 #include <asm/mach-imx/boot_mode.h>
 #include <asm/mach-imx/video.h>
+#include "../common/eeprom.h"
 
 DECLARE_GLOBAL_DATA_PTR;
+
+#define ENV_FDT_FILE "fdt_file"
+#define TYPE_2L_OPT "WBM" //Wi/Fi module is LBES5PL2EL by Murata, PCB rev 1.4
+static void ucm_imx8_mini_select_dtb(void)
+{
+	const char *fdt_src;
+	const char *env_fdt_file = env_get(ENV_FDT_FILE);
+	char buf[PRODUCT_OPTION_SIZE * PRODUCT_OPTION_NUM + 1];
+// Variable fdt_file unset or empty -- choose FDT, basing on the device ooptions
+	if (NULL == env_fdt_file || 0 == env_fdt_file[0]) {
+		env_fdt_file = CONFIG_DEFAULT_FDT_WB5; //Valid DTB for all PCB revisions, but 1.4
+		if (cl_eeprom_read_som_options(buf)) {
+			buf[-1] = 0; // Precaution, sinse we dont have strnstr function
+			if (strstr(buf, TYPE_2L_OPT))
+				env_fdt_file = CONFIG_DEFAULT_FDT_2EL;
+
+			fdt_src = "eeprom"; // board revision determines FDT name
+		} else { // Nothing found, m.b uninitialized eeprom
+			fdt_src = "default";
+		}
+		env_set(ENV_FDT_FILE, env_fdt_file);
+	} else {
+		fdt_src = "env"; // FDT name is gotten from environment
+	}
+
+	printf("FDT[%s]: %s %s\n", fdt_src, env_fdt_file);
+}
+
+int sub_board_late_init(void)
+{
+	ucm_imx8_mini_select_dtb();
+	return 0;
+}
+
+
diff --git a/configs/cl-imx8m-mini_defconfig b/configs/cl-imx8m-mini_defconfig
index f3c2d469eff..09c71cf7779 100644
--- a/configs/cl-imx8m-mini_defconfig
+++ b/configs/cl-imx8m-mini_defconfig
@@ -172,6 +172,7 @@ CONFIG_SPL_LIBCOMMON_SUPPORT=y
 CONFIG_SPL_LIBGENERIC_SUPPORT=y
 
 CONFIG_SYS_I2C_EEPROM_BUS=1
+CONFIG_SYS_I2C_EEPROM_ADDR=0x50
 CONFIG_SYS_EEPROM_PAGE_WRITE_BITS=4
 CONFIG_SYS_EEPROM_PAGE_WRITE_DELAY_MS=5
 CONFIG_SYS_I2C_MXC_I2C1=y
diff --git a/configs/ucm-imx8m-mini.config b/configs/ucm-imx8m-mini.config
index d9742031364..c0e0be19c82 100644
--- a/configs/ucm-imx8m-mini.config
+++ b/configs/ucm-imx8m-mini.config
@@ -1,4 +1,3 @@
 CONFIG_TARGET_UCM_IMX8M_MINI=y
 CONFIG_LOCALVERSION="-ucm-imx8m-mini-2.0"
 CONFIG_DEFAULT_DTB="ucm-imx8m-mini.dtb"
-CONFIG_DEFAULT_FDT_FILE="ucm-imx8m-mini.dtb"
-- 
2.34.1

